### PR TITLE
Refactored continuous distributions to v4 (AsymmetricLaplace, HalfStudentT, ExGaussian, Interpolated)

### PR DESCRIPTION
**Pull request by kc611**
Linked Issue: #4686 

This PR refactors the following distributions:

- AsymmetricLaplace
- HalfStudentT
-  ExGaussian
-  Interpolated

(Random Note: It's strange how git is showing that many file changes in the test file `pymc3/tests/test_distributions.py` when only difference I made is of few lines ?)

### CODE REVIEW

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [inline pymc3/distributions/continuous.py]: `lam` and `sd` are not being used, they should be removed, no?

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [inline pymc3/distributions/continuous.py]: Only one default parametrization should go into `super().dist`, probably `[nu, sigma]`. You can check the already refactored t-student for an example.

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [inline pymc3/distributions/continuous.py]: Similarly, the `logp` should receive only one default parametrization (the same you define in the `dist` classmethod.

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [inline pymc3/distributions/continuous.py]: Same here, `sd` should not be in here

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [inline pymc3/distributions/continuous.py]: the `interp_op` will not be reached from the`logp`, you will have to define it in there.

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [inline pymc3/distributions/continuous.py]: A bit confusingly, we are not using `self` instances, the logp has to work as a standard function dependent only on it's inputs. I am referring to the `self.interp_op`

**Review by ricardoV94 — CHANGES_REQUESTED** [review]
Thanks for opening the PR! I marked a couple of things than need to be changed. Let me know if you have any questions :)

Also, we will need the test in `test_distributions_random.py` to be refactored / added, and any tests that may be marked as `xfail` because of these distributions anywhere else on the test module.

File: pymc3/tests/test_distributions.py
Comment (ricardoV94) [inline pymc3/tests/test_distributions.py]: This new indentation is (probably) wrong

File: pymc3/tests/test_distributions.py
Comment (ricardoV94) [inline pymc3/tests/test_distributions.py]: This will need to be refactored for this PR, otherwise we have no test that covers the `logp` method of the `Interpolated`, or do we?

**Comment by ricardoV94** [comment]
> (Random Note: It's strange how git is showing that many file changes in the test file `pymc3/tests/test_distributions.py` when only difference I made is of few lines ?)

Could it be some auto-formatting on your end? I can try to check locally to see if it is just a GitHub confusion or if you actually indented those lines.

File: pymc3/tests/test_distributions.py
Comment (kc611) [reply pymc3/tests/test_distributions.py]: Yea, I plan on doing exactly that in this PR. It seems some problem with the arguments not being passed correctly.

**Comment by kc611** [comment]
> Could it be some auto-formatting on your end? I can try to check locally to see if it is just a GitHub confusion or if you actually indented those lines.

Please do. The only changes it shows on my side are the few changed `x_fail` lines.

Edit: Never mind, Looks like VS Code doesn't take into account the indentation changes as "real" changes in git. Fixed it now.

File: pymc3/distributions/continuous.py
Comment (kc611) [inline pymc3/distributions/continuous.py]: I'm a bit skeptical about this particular parametrization, since that required removing this condition over here.(I don't know if this particular condition is important or not, though)

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [inline pymc3/distributions/continuous.py]: everything that is not going into `dist` can be commented for the time being (and doesn't need the `as_tensor_variable` call either. This is related to sampler starting point logic that has not been refactored.

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [reply pymc3/distributions/continuous.py]: The lam sigma is a bit weird in that one can be converted from the other, even if the original is negative. You can obtain the lam again from sigma by  using the `get_tau_sigma` inside the `logp`. This way you can keep the lam check in the `bound`, but probably we should just open a separate issue to fix `get_tau_sigma`

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [inline pymc3/distributions/continuous.py]: mean and variance can be commented out for the same reason as below. They are not being used by the current codebase

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [inline pymc3/distributions/continuous.py]: typo `ssigma`

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [inline pymc3/distributions/continuous.py]: leftover typo: `mu = mu`

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [inline pymc3/distributions/continuous.py:3765]: seems like the rng_fn needs four arguments, so `ndims_params = [0, 0, 0, 0]`

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [inline pymc3/distributions/continuous.py]: Something is off with the `rng_fn` signature. `p` seems to be overwritten so it shouldn't be there, and if you want to pass the `cdf` already specified it should also be in the `dist` classmethod bellow and `ndim_params` shoud be 3 elements long.

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [inline pymc3/distributions/continuous.py]: I think `ndims_params` cannot be scalar (i.e., `0`), right? In that case it should be `ndims_params = [1, 1]`

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [inline pymc3/distributions/continuous.py:3889]: `Z` is not being passed to `super().dist()` and as such won't reach the `logp`. It seems it is not necessary for the random draws so it's probably better to also specify it inside the `logp` from the `x_points` input.

File: pymc3/distributions/continuous.py
Comment (kc611) [reply pymc3/distributions/continuous.py]: Ah, okay got it.

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [inline pymc3/distributions/continuous.py]: `ndims_params` should be `[0, 0]`, i.e., the same length as the number of parameters

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [inline pymc3/distributions/continuous.py]: No need to add the assert here, it should just be in the `dist` classmethod

File: pymc3/distributions/continuous.py
Comment (ricardoV94) [inline pymc3/distributions/continuous.py]: Inside `rng_fn` there shouldn't be any symbolic operations. I didn't realize you need `Z` here as well.

You can either use scipy equivalents or, if easier, have everything that's needed for both `logp` and the `rng_fn` defined once during the `dist classmethod` and passed to `super` as additional inputs (and expected in both the `logp` and `rng_fn`. This way, the `rng_fn` will receive the non-symbolic arguments as inputs during normal sampling, while the `logp` method will receive the `symbolic` ones. 

Apologies if this sounds very confusing :D

Comment (kc611) [reply pymc3/distributions/continuous.py]: I think I should probably refactor the test first, just so that I'll have an intuition of how this entire `Interpolated` thing should work.

Comment (ricardoV94) [reply pymc3/distributions/continuous.py]: Yeah, you can also check it out in V3, and see how it works there (and in the test). That may help, since this is a very non-standard one

File: pymc3/distributions/continuous.py
Comment (kc611) [inline pymc3/distributions/continuous.py:3806]: It looks like we cannot directly access the index of `pdf` or `x` like this. (The `index` variable is an `Elemwise` during the initial graph construction and hence this fails.). What class/function will be a good substitute for this?

Comment (ricardoV94) [reply pymc3/distributions/continuous.py:3806]: I think the problem is that you are using symbolic operations inside the rng_fn. Everything there should be non-symbolic AFAIK

Comment (kc611) [reply pymc3/distributions/continuous.py:3806]: That indeed is the problem. What I wanted to know was what is the non-symbolic substitute for such operation?

 (If there isn't I'll try your suggestion from https://github.com/pymc-devs/pymc3/pull/4746#discussion_r646490859 )

Comment (ricardoV94) [reply pymc3/distributions/continuous.py:3806]: I think the symbolic ops used in here are all wrapping standard scipy function: https://github.com/pymc-devs/pymc3/blob/25eaa715018ed7add3ba1289b6ad068636e29e5e/pymc3/distributions/dist_math.py#L318

Comment (kc611) [reply pymc3/distributions/continuous.py:3806]: I tried fiddling around with `SplineWrapper` but I don't think that'll be of much use here since we are using different methods of `InterpolatedUnivariateSpline` like the `.integral()` and the `.antiderivate()` ( But can we actually use `SplineWrapper` like that ? )

Instead as you mentioned in https://github.com/pymc-devs/pymc3/pull/4746#discussion_r646490859 I just kept the non-symbolic operations in `dist` and forwarded `cdf_points` as an extra argument.

**Review by michaelosthege — APPROVED** [review]
LGTM
@ricardoV94 you make the call

File: pymc3/tests/test_distributions_random.py
Comment (ricardoV94) [inline pymc3/tests/test_distributions_random.py]: should the import be at the top?

File: pymc3/tests/test_distributions_random.py
Comment (ricardoV94) [inline pymc3/tests/test_distributions_random.py]: These tests that rely on `BaseTestCases.BaseTestCase` should be removed and replaced with tests using the new `BaseTestDistribution`

**Review by ricardoV94 — CHANGES_REQUESTED** [review]
Looks great so far. The random tests still need to be refactored to use the `BaseTestDistribution`

File: pymc3/tests/test_distributions_random.py
Comment (ricardoV94) [inline pymc3/tests/test_distributions_random.py]: This one should be moved to it's own class. The reason for this is that after refactoring all distributions, the `TestScalarParameterSamples` where this one is will disappear, but this distribution is special enough that the test should be kept (i.e., we are not simply drawing from scipy / numpy distributions).

File: pymc3/tests/test_distributions_random.py
Comment (ricardoV94) [inline pymc3/tests/test_distributions_random.py:1260]: These tests that are using a `ref_rand` should be removed as well, unless there is a good reason to test them against a reference. If we are just copying the same scipy/numpy expression we use in our `random Ops` the default tests in `BaseTestDistribution` should be used instead.

File: pymc3/tests/test_distributions_random.py
Comment (kc611) [reply pymc3/tests/test_distributions_random.py]: I just moved it to it's own class and named it `TestInterpolated`

**Review by ricardoV94 — COMMENTED** [review]
Looks good. The last thing I would suggest is to add a `BaseTestDistribution` for the `Interpolated` just with the `check_rv_size` test enabled. Just to make sure `size` is working properly.

Otherwise, I think this PR is pretty much spot on. Specially wth the thorny `Interpolated`.

**Comment by kc611** [comment]
Thanks @ricardoV94 for all your suggestions, (and especially for being patient with me in the start of this PR with all those silly mistakes. I guess my coffee didn't work quite right that particular day.)

Edit: Also (just a heads up) I'll be working on refactoring `KroneckerNormal` from the multivariate distributions next.

**Comment by ricardoV94** [comment]
I am afraid the `Interpolated` is still problematic. Right now it works fine if you pass non-symbolic arguments, but it fails if the inputs are symbolic. This snippet raises an enigmatic `ValueError` from within the `scipy.interpolate` method:

```python
x_points = np.linspace(0, 10)
pdf_points = st.norm.pdf(x_points, loc=5, scale=0.1)
ae_x_points = at.as_tensor_variable(x_points)
ae_pdf_points = at.as_tensor_variable(pdf_points)
interp = pm.Interpolated.dist(ae_x_points, ae_pdf_points, size=(2, 4))
# ValueError: diff requires input that is at least one dimensional
```

`V3` also had the same limitation.

It seems a proper fix would require integrating the non-symbolic `InterpolatedUnivariateSpline` inside the symbolic `SplineWrapper` (or as a separate symbolic Op). The `logp` method making use of `x_points.data` is also quite hackish. For instance, it wouldn't work with a generic `at.arange` input.

~~If we can't be bothered at the moment, perhaps we should at least mention this limitation explicitly in the docstring and add an informative `TypeError` in `dist` if the input is symbolic. What do you think?~~

Edit: I see this line is in the docstrings:
```
Both parameters ``x_points`` and values ``pdf_points`` are not variables, but
plain array-like objects, so they are constant and cannot be sampled.
```

**Comment by ricardoV94** [comment]
I pushed some changes that hopefully clarify a bit how the `Interpolated` is expected to work internally, ~~and also placed the contents of the `_argcdf` method directly inside the `rng_fn`.~~

I am okay with merging as is, and perhaps open a GH issue to discuss whether we want to make this distribution accept symbolic elements in the future.

**Review by ricardoV94 — APPROVED** [review]
Great work @kc611 and thanks for your patience. Let me know if you also agree with merging it, given my last comments.

**Comment by kc611** [comment]
Sure, you can go ahead and merge this, and yes, we should probably open an issue for making a proper fix for integrating the non-symbolic `InterpolatedUnivariateSpline` inside the symbolic `SplineWrapper` (or as a separate symbolic Op). Otherwise we'll be stuck with this hackish approach for `Interpolated` which I'm sure we don't want for `v4`.

**Comment by ricardoV94** [comment]
> Sure, you can go ahead and merge this, and yes, we should probably open an issue for making a proper fix for integrating the non-symbolic `InterpolatedUnivariateSpline` inside the symbolic `SplineWrapper` (or as a separate symbolic Op). Otherwise we'll be stuck with this hackish approach for `Interpolated` which I'm sure we don't want for `v4`. 

Do you mind creating that issue?

I'll wait until Monday to merge the PR in case anyone else wants to have a look.

**Pull request merged**
